odl/tomo/backends/astra_cuda.py

```diff
@@ -38,207 +38,340 @@
-    if not isinstance(geometry, Geometry):
-        raise TypeError('geometry  {!r} is not a Geometry instance'
-                        ''.format(geometry))
-    if vol_data.ndim != geometry.ndim:
-        raise ValueError('dimensions {} of volume data and {} of geometry '
-                         'do not match'
-                         ''.format(vol_data.ndim, geometry.ndim))
-    if not isinstance(proj_space, DiscreteLp):
-        raise TypeError('projection space {!r} is not a DiscreteLp '
-                        'instance'.format(proj_space))
-    if out is not None:
-        if not isinstance(out, DiscreteLpElement):
-            raise TypeError('`out` {} is neither None nor a '
-                            'DiscreteLpElement instance'.format(out))
-
-    ndim = vol_data.ndim
-
-    # Create astra geometries
-    vol_geom = astra_volume_geometry(vol_data.space)
-    proj_geom = astra_projection_geometry(geometry)
-
-    # Create ASTRA data structures
-
-    # In the case dim == 3, we need to swap axes, so can't perform the FP
-    # in-place
-    vol_id = astra_data(vol_geom, datatype='volume', data=vol_data,
-                        allow_copy=True)
-
-    # Create projector
-    proj_id = astra_projector('nearest', vol_geom, proj_geom, ndim,
-                              impl='cuda')
-
-    if ndim == 2:
-        if out is None:
-            out = proj_space.element()
+    def __init__(self, geometry, reco_space, proj_space, use_cache):
+        """Run an ASTRA forward projection on the given data using the GPU.
 
-        # Wrap the array in correct dtype etc if needed
-        with writable_array(out, dtype='float32', order='C') as arr:
-            sino_id = astra_data(proj_geom, datatype='projection', data=arr)
+        Parameters
+        ----------
+        geometry : `Geometry`
+            Geometry defining the tomographic setup
+        out : ``proj_space`` element, optional
+            Element of the projection space to which the result is written. If
+            ``None``, an element in ``proj_space`` is created.
+        use_cache : bool
+            True if data should be cached. Default: True
+        """
+        if not isinstance(geometry, Geometry):
+            raise TypeError('geometry  {!r} is not a Geometry instance'
+                            ''.format(geometry))
+        self.geometry = geometry
+
+        if not isinstance(reco_space, DiscreteLp):
+            raise TypeError('reconstruciton space {!r} is not a DiscreteLp '
+                            'instance'.format(proj_space))
+        self.reco_space = reco_space
+
+        if not isinstance(proj_space, DiscreteLp):
+            raise TypeError('projection space {!r} is not a DiscreteLp '
+                            'instance'.format(proj_space))
+        self.proj_space = proj_space
+
+        self.use_cache = bool(use_cache)
+
+        if self.use_cache:
+            self.create_ids()
+
+    def call_forward(self, vol_data, out=None):
```

Comment:
I had a plan of having a single object for both the forward and back projection, that way we get half the memory overhead. Naming like this is a step for enabling that later.